### PR TITLE
changed user table to student and added login and register tests

### PR DESCRIPTION
**Pull request by EyalAzimov**

### CODE REVIEW

**Comment by based-kwl** [comment]
Just a note: these http tests to mongo won't run properly until we have a public API available.

**Comment by guaj** [comment]
> Just a note: these http tests to mongo won't run properly until we have a public API available. So the CircleCI tests failing is expected.
> 
> I have ran the tests locally and they work when the server is running.

Hey, sorry just wen through this comment. The ec2 isntance went down. Looking to boot it up. Althought, jest test should work out of the box, meaning that they should be able to run without any data base connections. Make sure you are testing the right thing.

**Comment by based-kwl** [comment]
> > Just a note: these http tests to mongo won't run properly until we have a public API available. So the CircleCI tests failing is expected.
> > I have ran the tests locally and they work when the server is running.
> 
> Hey, sorry just wen through this comment. The ec2 isntance went down. Looking to boot it up. Althought, jest test should work out of the box, meaning that they should be able to run without any data base connections. Make sure you are testing the right thing.

This is not what i meant. The tests in question do work if you have the secret key, but since connecting to the database requires a secret key which is not included in the repo, CircleCI will always fail. It doesn't mean that the tests are wrong, it's simply because we are using MongoDB atlas, a cloud db and not a local db. 

Once we have a permanent backend server up and running on a 24/7 server with an accessible api (e.g. https://www.semester-planner.com/api/student/ instead of localhost:5000), these test will function as intended after changing the http request address to that of the API.

**Comment by guaj** [comment]
> > > Just a note: these http tests to mongo won't run properly until we have a public API available. So the CircleCI tests failing is expected.
> > > I have ran the tests locally and they work when the server is running.
> > 
> > 
> > Hey, sorry just wen through this comment. The ec2 isntance went down. Looking to boot it up. Althought, jest test should work out of the box, meaning that they should be able to run without any data base connections. Make sure you are testing the right thing.
> 
> This is not what i meant. The tests in question do work if you have the secret key, i.e. if you run the test locally, but since the key is not included in the repo (for obvious security reasons), CircleCI will always fail because it cannot make those http requests and the tests will fail. It doesn't mean that the tests are wrong, it's simply because we are using MongoDB atlas, a cloud db and not a local db.
> 
> Once we have a permanent backend server up and running on a 24/7 server with an accessible api (e.g. https://www.semester-planner.com/api/student/ instead of localhost:5000), these test will function as intended after changing the http request address to that of the API.

these tests don't work because env variables are not passed into circle ci. Im fixing it and ill write a test for the mongodb connection

File: __test__/Student_test_data.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+
+registerPayload ={"username":"test45",
+"password":"scooby",
+"faculty":"encs",
+"email":"ram@b.ca",
+"program":"coen",
+"privateProfile":"true",
+"password":"scooby"}
+
```
Comment (guaj) [inline __test__/Student_test_data.js:9]: this should be passed as an env varible. Ill add them to the circle ci env

File: src/routes/Student.js [hunk 2]
```diff
@@ -49,8 +49,17 @@ router.route('/email/:email').get((req, res) => {
  * Delete a user by ID
  */
 router.route('/:id').delete((req, res) => {
-    User.findByIdAndDelete(req.params.id)
```
Comment (guaj) [inline src/routes/Student.js:52]: we don't need both, just the email route is fine

File: src/routes/Student.js [hunk 3]
```diff
@@ -104,7 +113,7 @@ router.route('/add').post( async (req, res) => {
     const faculty = req.body.faculty;
     const privateProfile= req.body.privateProfile;
 
-    const newUser = new User({
+    const newStudent = new Student({
```
Comment (guaj) [inline src/routes/Student.js:116]: student creation responsability should not be in the route. We can define a student factory in another file that will create students based on the data passed in the request. We want low coupling here.

File: src/routes/login.js [hunk 4]
```diff
@@ -1,5 +1,5 @@
 const router = require('express').Router();
-const User = require("../models/user.model");
+const Student = require("../models/Student.model");
```
Comment (guaj) [inline src/routes/login.js:2]: constant usually starts with lower letters not upper ones. No consistency in the files.

File: __test__/Student.spec.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
+const {registerPayload, loginPayload, deletePayload} =  require("./Student_test_data");
+const axios = require("axios");
+
+//These tests will work locally, but not in CicleCI because we do not have a public API available.
+//Change the URL when that is setup.
+test("add a new Student ", async () => {
+  
+  let httpResponse = await axios.post('http://localhost:5000/Student/add',registerPayload)
```
Comment (guaj) [inline __test__/Student.spec.js:8]: Do not use axios here ... import the route directly and use it.
see here  : https://fek.io/blog/how-to-add-unit-testing-to-express-using-jest

File: __test__/Student.spec.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    .then( response =>{
+          return response.data
+    });
+  expect(httpResponse).toBe("Student ram@b.ca added");
+ });
+
+
+test("student login test ", async () => {
+  
+  let httpResponse = await axios.post('http://localhost:5000/login', loginPayload)
```
Comment (guaj) [inline __test__/Student.spec.js:18]: same

File: __test__/Student.spec.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    });
+
+  expect(httpResponse).toBe("test45");
+ 
+});
+
+//works as cleanup of previous test too.
+test("delete a Student ", async () => {
+  
+  let httpResponse = await axios.delete('http://localhost:5000/student/email/ram@b.ca')
```
Comment (guaj) [inline __test__/Student.spec.js:31]: same

**Review by guaj — CHANGES_REQUESTED** [review]

**Comment by based-kwl** [comment]
I am legitimately not sure what you are trying to do. I'm telling you CircleCI doesn't have our secret key to connect to mongo... yet you keep trying to make it connect via other means. 

If you want to have the same environment that CircleCI has, delete your .env file and try to pass the test...

**Comment by guaj** [comment]
> I am legitimately not sure what you are trying to do. I'm telling you CircleCI doesn't have our secret key to connect to mongo... yet you keep trying to make it connect via other means.
> 
> If you want to have the same environment that CircleCI has, delete your .env file and try to pass the test...

circle ci has the secrents ... we can discuss this in the meeting tonight.

**Comment by based-kwl** [comment]
> > I am legitimately not sure what you are trying to do. I'm telling you CircleCI doesn't have our secret key to connect to mongo... yet you keep trying to make it connect via other means.
> > If you want to have the same environment that CircleCI has, delete your .env file and try to pass the test...
> 
> circle ci has the secrets ... my test (mongoDbConnection.spec.ts) pass in the pipeline. We can discuss this in the meeting tonight.

Yeah you were right, changed the tests to not use axios and fixed some consistency issues.

**Comment by RamB98** [comment]
pull request reviewed  all test have passed, merging the branch

**Pull request merged**